tfe/provider.go

```diff
@@ -0,0 +1,200 @@
+			"hostname": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: descriptions["hostname"],
+				Default:     defaultHostname,
+			},
+
+			"token": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: descriptions["token"],
+			},
+		},
+
+		ResourcesMap: map[string]*schema.Resource{
+			"tfe_organization":       resourceTFEOrganization(),
+			"tfe_organization_token": resourceTFEOrganizationToken(),
+			"tfe_sentinel_policy":    resourceTFESentinelPolicy(),
+			"tfe_ssh_key":            resourceTFESSHKey(),
+			"tfe_team":               resourceTFETeam(),
+			"tfe_team_access":        resourceTFETeamAccess(),
+			"tfe_team_member":        resourceTFETeamMember(),
+			"tfe_team_members":       resourceTFETeamMembers(),
+			"tfe_team_token":         resourceTFETeamToken(),
+			"tfe_workspace":          resourceTFEWorkspace(),
+			"tfe_variable":           resourceTFEVariable(),
+		},
+
+		ConfigureFunc: providerConfigure,
+	}
+}
+
+func providerConfigure(d *schema.ResourceData) (interface{}, error) {
+	// Get the hostname and token.
+	hostname := d.Get("hostname").(string)
+	token := d.Get("token").(string)
+
+	// Parse the hostname for comparison,
+	host, err := svchost.ForComparison(hostname)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the Terraform CLI configuration.
+	config := cliConfig()
+
+	// Create a new credential source and service discovery object.
+	credsSrc := credentialsSource(config)
+	services := disco.NewWithCredentialsSource(credsSrc)
+
+	// Add any static host configurations service discovery object.
+	for userHost, hostConfig := range config.Hosts {
+		host, err := svchost.ForComparison(userHost)
+		if err != nil {
+			// ignore invalid hostnames.
+			continue
+		}
+		services.ForceHostServices(host, hostConfig.Services)
+	}
+
+	// Discover the full Terraform Enterprise service address.
+	address := services.DiscoverServiceURL(host, serviceID)
+	if address == nil {
+		return nil, fmt.Errorf("host %s does not provide a Terraform Enterprise API", host)
+	}
+
+	// Only try to get to the token from the credentials source if no token
+	// was explicitly set in the provider configuration.
+	if token == "" {
+		creds, err := services.CredentialsForHost(host)
+		if err != nil {
+			log.Printf("[DEBUG] Failed to get credentials for %s: %s (ignoring)", host, err)
+		}
+		if creds != nil {
+			token = creds.Token()
+		}
+	}
+
+	// If we still don't have a token at this point, we return an error.
+	if token == "" {
+		return nil, fmt.Errorf("required token could not be found")
+	}
+
+	// Create a new TFE client config..
+	cfg := &tfe.Config{
+		Address: address.String(),
+		Token:   token,
+	}
+
+	// Create s new TFE client.
+	return tfe.NewClient(cfg)
+}
+
+// cliConfig tries to find and parse the configuration of the Terraform CLI.
+// This is an optional step, so any errors are ignored.
+func cliConfig() *Config {
```

Comment:
Ideally, this would return an error, instead of swallowing them.

Reply:
Yeah, I can see your point... Yet again this logic is copied over from Terraform itself. Terraform handles this logic in the exact same way. Now I'm not saying it's better because of that, but I just think this part should probably work the same so it delivers the same UX.

Reply:
I think where content is copied from elsewhere, a comment identifying that within the code would be ideal, just so we can trace the thought process in the future. I'm not sure tying this to Terraform's way of doing things makes sense, or if Terraform should be updated, or both, or neither. But I'm also not convinced we have to get to the bottom of that today, so for the moment, I think a comment would suffice.

Reply:
Agree. Will update when we release a new version!